File: Behavioral/State/typescript/state.ts
export namespace StatePattern {
   abstract class State {
      abstract handle(context: Context): void;
   }

   class ConcreteStateA implements State {
      public handle(context: Context): void {
         console.log("Handle method for: ConcreteStateA");
         context.State = new ConcreteStateB();
      }
   }

   class ConcreteStateB implements State {
      public handle(context: Context): void {
         console.log("Handle method for: ConcreteStateB");
         context.State = new ConcreteStateC();
      }
   }

   class ConcreteStateC implements State {
      public handle(context: Context): void {
         console.log("Final state");
         context.State = this;
      }
   }

   class Context {
      private state: State;

      constructor(state: State) {
         this.state = state;
      }

      get State(): State {
         return this.state;
      }

      set State(state: State) {
         this.state = state;
      }

      public request(): void {
         console.log("request is being called!");
         this.state.handle(this);
      }
   }

   export function demo(): void {
      const concreteStateA = new ConcreteStateA();
      const context: Context = new Context(concreteStateA);

      context.request();
      context.request();
      context.request(); // infinite loop to final state
   }
}

File: Behavioral/State/kotlin/State.kt
sealed class State

class ConcreteStateA() : State()

class ConcreteStateB() : State()

class Context {
    private var state : State = ConcreteStateA()

    fun setStateB() {
        state = ConcreteStateB()
    }

    fun setStateA() {
        state = ConcreteStateA()
    }


    val stateText : String
        get() = when(state) {
            is ConcreteStateA -> "State A"
            is ConcreteStateB -> "State B"
        }

    override fun toString() = "State: $stateText"
}

fun main(args: Array<String>) {
    val context = Context()
    context.setStateA()
    println(context)
    context.setStateB()
    println(context)
}